__tests__/errors.spec.ts

```diff
@@ -0,0 +1,347 @@
+        errors: [
+          {
+            path: 'nested',
+            error: {
+              expected: 'object',
+              got: { deeper: { invalidDate: 'I really like this library!' } },
+              errors: [
+                {
+                  path: 'deeper',
+                  error: {
+                    expected: 'object',
+                    got: { invalidDate: 'I really like this library!' },
+                    errors: [
+                      {
+                        path: 'invalidDate',
+                        error: { expected: 'date', got: 'I really like this library!' },
+                      },
+                    ],
+                  },
+                },
+                { path: 'invalidString', error: { expected: 'string', got: undefined } },
+              ],
+            },
+          },
+          { path: 'invalidNumber', error: { expected: 'number', got: undefined } },
+        ],
+      },
+    );
+  });
+
+  it('should list nested objects which are undefined', () => {
+    const validator = pipe(
+      object({
+        nested: object({
+          invalid: string(),
+        })(),
+      }),
+      validate,
+    );
+
+    expectToMatchError(() => validator({}), {
+      expected: 'object',
+      got: {},
+      errors: [{ path: 'nested', error: { expected: 'object', got: undefined } }],
+    });
+  });
+
+  it('should list nested objects which are defined', () => {
+    const validator = pipe(
+      object({
+        a: number(),
+        nested: object({
+          invalid: string(),
+        })(),
+      }),
+      validate,
+    );
+
+    expectToMatchError(() => validator({ a: 123 }), {
+      expected: 'object',
+      got: { a: 123 },
+      errors: [{ path: 'nested', error: { expected: 'object', got: undefined } }],
+    });
+  });
+
+  it('should use custom refinement to string', () => {
+    const email = refine<string, string>(
+      (value: string, t) =>
+        value.includes('@') ? t.nextValid(value) : t.left({ expected: 'email', got: value }),
+      () => 'email',
+    );
+
+    const validator = pipe(
+      object({
+        shouldBeEmail: string(email()),
+      }),
+      validate,
+    );
+
+    expectToMatchError(() => validator({ shouldBeEmail: 123 }), {
+      expected: 'object',
+      got: { shouldBeEmail: 123 },
+      errors: [
+        {
+          path: 'shouldBeEmail',
+          error: {
+            expected: 'string',
+            got: 123,
+          },
+        },
+      ],
+    });
+
+    expectToMatchError(() => validator({ shouldBeEmail: 'notanemail' }), {
+      expected: 'object',
+      got: { shouldBeEmail: 123 },
+      errors: [
+        {
+          path: 'shouldBeEmail',
+          error: {
+            expected: 'email',
+            got: 'notanemail',
+          },
+        },
+      ],
+    });
+  });
+
+  it('nullable', () => {
+    expectToMatchError(() => λ(string, nullable, validate)(123), {
+      expected: ['string', 'null'],
```

Comment:
@mmiszy I do not understand what value it brings. `nullable` or `optional` basically can not fail and this information here is useless from my point of view. I will revert it to make it only a string again but we can discuss it further. Can you point me to a good use case of this one?